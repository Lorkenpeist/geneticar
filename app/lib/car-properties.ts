// A car property that can be configured via sliders and/or genes
export interface CarPropSpec {
  label: string;
  min: number;
  max: number;
  default: number;
  // The step size for discrete properties, if applicable
  // If not specified, the property is considered continuous
  stepSize?: number;
}

export const CAR_PROP_KEYS = [
  "length",
  "height",
  "wheelRadius",
  "engineTorque",
  "rpmLimit",
] as const;
type CarPropKey = (typeof CAR_PROP_KEYS)[number];

export const CAR_PROP_SPECS: Record<CarPropKey, CarPropSpec> = {
  length: {
    label: "Length",
    min: 10,
    max: 100,
    default: 100,
  },
  height: {
    label: "Height",
    min: 10,
    max: 100,
    default: 50,
  },
  wheelRadius: {
    label: "Wheel Radius",
    min: 10,
    max: 50,
    default: 30,
  },
  engineTorque: {
    label: "Engine Torque",
    min: 0,
    max: 1,
    default: 0.5,
  },
  rpmLimit: {
    label: "RPM Limit",
    min: 100,
    max: 1000,
    default: 300,
  },
};

export type CarProperties = Record<CarPropKey, number>;
