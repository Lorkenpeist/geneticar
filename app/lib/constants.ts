// Constants for the simulation options
export const CAR_COUNT_MIN = 1;
export const CAR_COUNT_MAX = 4;
export const CAR_COUNT_DEFAULT = 3;
export const CAR_COUNT_LABEL = "Number of Cars";

export const ROAD_SEGMENT_LENGTH = 100;
export const ROAD_SEGMENT_HEIGHT = 10;

// Constants for rendering
export const WHEEL_SPRITE_LOCATION = "/wheel-sectors-svgrepo-com.svg";
export const WHEEL_SPRITE_SIZE = 800;
