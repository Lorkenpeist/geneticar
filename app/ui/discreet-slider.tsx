export default function DiscreetSlider({
  id,
  name,
  label,
  min,
  max,
  step = 1,
  defaultValue,
}: {
  id: string;
  name: string;
  label: string;
  min: number;
  max: number;
  step?: number;
  defaultValue: number;
}) {
  return (
    <div className="flex flex-col items-center text-xs font-medium">
      <label htmlFor={id} className="ml-2 flex">
        {label}
      </label>
      <div className="flex flex-row space-x-1 items-center">
        <div>{min}</div>
        <input
          id={id}
          name={name}
          type="range"
          min={min}
          max={max}
          step={step}
          defaultValue={defaultValue}
          list={`${id}:markers`}
          className="cursor-pointer"
        />
        <datalist id={`${id}:markers`}>
          {Array.from(
            { length: (max - min) / step + 1 },
            (_, i) => i * step + min,
          ).map((i) => (
            <option value={i} key={i.toString()}></option>
          ))}
        </datalist>
        <div>{max}</div>
      </div>
    </div>
  );
}
