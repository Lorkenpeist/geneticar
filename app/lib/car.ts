import { Body, Bodies, Composite, Constraint, Vector } from "matter-js";
import { WHEEL_SPRITE_LOCATION, WHEEL_SPRITE_SIZE } from "./constants";

export interface CarProperties {
  width: number;
  height: number;
  wheelSize: number;
  torque: number;
}

export class Car {
  // The physical properties of the car
  readonly props: CarProperties;
  // The physical Matter.js representation of the car
  readonly composite: Composite;
  // the body of the car
  private readonly body: Body;
  // The wheels of the car
  private readonly wheels: Body[];

  constructor(props: CarProperties, position: Vector) {
    this.props = { ...props };
    const group = Body.nextGroup(true);
    const wheelAOffset = -props.width / 2;
    const wheelBOffset = props.width / 2;
    const wheelYOffset = props.height / 2;

    this.composite = Composite.create({ label: "Car" });

    this.body = Bodies.rectangle(0, 0, props.width, props.height, {
      collisionFilter: { group: group },
      density: 0.0002,
    });

    this.wheels = [
      this.wheel(group, wheelAOffset, wheelYOffset),
      this.wheel(group, wheelBOffset, wheelYOffset),
    ];

    const axels = this.wheels.map((wheel) =>
      Constraint.create({
        bodyA: this.body,
        pointA: { x: wheel.position.x, y: wheel.position.y },
        bodyB: wheel,
      }),
    );

    Composite.add(this.composite, this.body);
    Composite.add(this.composite, this.wheels);
    Composite.add(this.composite, axels);

    // calculate the position where the center of the body should be placed.
    // The provided position is the x coordinate of the frontof the car
    // and the y coordinate of the bottom of the wheels.
    const front = Math.max(
      ...this.composite.bodies
        .flatMap((body) => body.vertices)
        .map((vertex) => vertex.x),
    );

    // Yes, max, because higher y value is lower position.
    // Don't ask why.
    const bottom = Math.max(
      ...this.composite.bodies
        .flatMap((body) => body.vertices)
        .map((vertex) => vertex.y),
    );

    Composite.translate(this.composite, {
      x: position.x - front,
      y: position.y - bottom,
    });
  }

  // apply torque to the wheels
  drive() {
    this.wheels.forEach((wheel) => {
      // TODO: variable torque balance
      wheel.torque += this.props.torque / 2;
      console.log(`My torque is ${this.props.torque}`);
    });
  }

  // construct the wheel
  private wheel(group: number, x: number, y: number) {
    return Bodies.circle(x, y, this.props.wheelSize, {
      collisionFilter: { group },
      friction: 0.8,
      render: {
        sprite: {
          texture: WHEEL_SPRITE_LOCATION,
          xScale: (2 * this.props.wheelSize) / WHEEL_SPRITE_SIZE,
          yScale: (2 * this.props.wheelSize) / WHEEL_SPRITE_SIZE,
        },
      },
    });
  }
}
