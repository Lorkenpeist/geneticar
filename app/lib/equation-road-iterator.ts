import { Bodies, Vector } from "matter-js";
import { ROAD_SEGMENT_HEIGHT, ROAD_SEGMENT_LENGTH } from "./constants";

// Iteratively generates segments of road from a mathematical equation.
// Segments are generated to approximate the shape of the curve defined by the equation.
export function* equationRoadGenerator(f: (x: number) => number) {
  let x = 0;
  let y = f(0);
  while (true) {
    const x2 = x + ROAD_SEGMENT_LENGTH;
    const y2 = f(x2);
    yield segment({ x, y }, { x: x2, y: y2 });
    x = x2;
    y = y2;
  }
}

// Creates a road segment from 2 vertices by adding 2 more vertices below them
function segment(vertex1: Vector, vertex2: Vector) {
  const vertex3 = { x: vertex1.x, y: vertex1.y + ROAD_SEGMENT_HEIGHT };
  const vertex4 = { x: vertex2.x, y: vertex2.y + ROAD_SEGMENT_HEIGHT };

  return Bodies.fromVertices(
    // We have to calculate the center because Matter.js will automatically
    // move the center of mass to the provided coordinates
    (vertex1.x + vertex2.x) / 2,
    (vertex1.y + vertex2.y + ROAD_SEGMENT_HEIGHT) / 2,
    [[vertex1, vertex2, vertex3, vertex4]],
    { isStatic: true },
  );
}
