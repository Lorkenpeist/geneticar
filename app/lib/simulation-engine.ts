import { Engine, Runner, Bodies, Composite, Events, World } from "matter-js";
import { Car, CarProperties } from "./car";

export interface SimulationOptions {
  carCount: number;
  carProperties: CarProperties;
}

class SimulationEngine {
  readonly engine = Engine.create();
  private readonly runner = Runner.create();
  private cars: Car[] = [];

  start(options: SimulationOptions) {
    this.cars = Array.from(Array(options.carCount), (_, i) => {
      const car = new Car(options.carProperties, {
        x: 200 + 200 * i,
        y: 580,
      });
      return car;
    });
    const ground = Bodies.rectangle(400, 610, 810, 60, { isStatic: true });

    Composite.add(this.engine.world, [
      ...this.cars.map((car) => car.composite),
      ground,
    ]);

    Events.on(this.engine, "beforeUpdate", () =>
      this.cars.forEach((car) => car.drive()),
    );

    Runner.run(this.runner, this.engine);
  }

  stop() {
    // FIXME: update this after fixing Matter.js type definitions
    // @ts-expect-error Expected 3 arguments, but got 2.
    Events.off(this.engine, "beforeUpdate");
    Runner.stop(this.runner);
    Engine.clear(this.engine);
    World.clear(this.engine.world, true);
    this.cars = [];
  }
}

export const simulationEngine = new SimulationEngine();
