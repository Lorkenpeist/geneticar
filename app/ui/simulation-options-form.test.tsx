import { fireEvent, render } from "@testing-library/react";
import "@testing-library/jest-dom";
import SimulationOptionsForm from "./simulation-options-form";
import {
  CAR_COUNT_DEFAULT,
  CAR_COUNT_LABEL,
  CAR_COUNT_MAX,
} from "../lib/constants";
import { DiscreetSliderProps } from "./discreet-slider";
import { CAR_PROP_SPECS } from "../lib/car-properties";

jest.mock("./discreet-slider", () => ({
  __esModule: true,
  default: ({
    id,
    name,
    label,
    min,
    max,
    step = 1,
    defaultValue,
  }: DiscreetSliderProps) => (
    <input
      id={id}
      name={name}
      type="range"
      min={min}
      max={max}
      step={step}
      defaultValue={defaultValue}
      aria-label={label}
    />
  ),
}));

describe(SimulationOptionsForm.name, () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const simulationOptions = {
    carCount: CAR_COUNT_DEFAULT,
    carProperties: {
      length: CAR_PROP_SPECS.length.default,
      height: CAR_PROP_SPECS.height.default,
      wheelRadius: CAR_PROP_SPECS.wheelRadius.default,
      engineTorque: CAR_PROP_SPECS.engineTorque.default,
      rpmLimit: CAR_PROP_SPECS.rpmLimit.default,
    },
  };
  const setSimulationOptions = jest.fn();

  const simulationOptionsForm = (
    <SimulationOptionsForm
      options={simulationOptions}
      setOptions={setSimulationOptions}
    />
  );

  it("renders correctly", () => {
    const tree = render(simulationOptionsForm);
    expect(tree.container).toMatchSnapshot();
  });

  it("sets simulation options when form is submitted", async () => {
    const tree = render(simulationOptionsForm);

    fireEvent.change(tree.getByLabelText(CAR_COUNT_LABEL), {
      target: { value: CAR_COUNT_MAX },
    });

    Object.values(CAR_PROP_SPECS).forEach((spec) => {
      fireEvent.change(tree.getByLabelText(spec.label), {
        target: { value: spec.max },
      });
    });

    fireEvent.submit(tree.container.querySelector("form")!);

    expect(setSimulationOptions).toHaveBeenCalledTimes(1);
    expect(setSimulationOptions).toHaveBeenCalledWith({
      carCount: CAR_COUNT_MAX,
      carProperties: Object.fromEntries(
        Object.entries(CAR_PROP_SPECS).map(([key, spec]) => [key, spec.max]),
      ),
    });
  });
});
