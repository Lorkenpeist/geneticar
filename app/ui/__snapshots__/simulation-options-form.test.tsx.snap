// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimulationOptionsForm renders correctly 1`] = `
<div>
  <form
    action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
  >
    <div
      class="rounded-md bg-gray-100 dark:bg-gray-800 p-4 md:p-6"
    >
      <fieldset>
        <legend
          class="mb-2 text-sm font-medium"
        >
          Simulation Options
        </legend>
        <div
          class="rounded-md border border-gray-200 bg-white dark:bg-gray-900 px-[14px] py-3"
        >
          <div
            class="flex flex-col gap-4"
          >
            <input
              aria-label="Number of Cars"
              id="carCount"
              max="4"
              min="1"
              name="carCount"
              step="1"
              type="range"
              value="3"
            />
            <input
              aria-label="Length"
              id="car_length"
              max="100"
              min="10"
              name="car_length"
              step="10"
              type="range"
              value="100"
            />
            <input
              aria-label="Height"
              id="car_height"
              max="100"
              min="10"
              name="car_height"
              step="10"
              type="range"
              value="50"
            />
            <input
              aria-label="Wheel Radius"
              id="car_wheelRadius"
              max="50"
              min="10"
              name="car_wheelRadius"
              step="10"
              type="range"
              value="30"
            />
            <input
              aria-label="Engine Torque"
              id="car_engineTorque"
              max="1"
              min="0"
              name="car_engineTorque"
              step="0.1"
              type="range"
              value="0.5"
            />
            <input
              aria-label="RPM Limit"
              id="car_rpmLimit"
              max="1000"
              min="100"
              name="car_rpmLimit"
              step="100"
              type="range"
              value="300"
            />
          </div>
        </div>
      </fieldset>
      <div
        class="mt-6 flex justify-end gap-4"
      >
        <button
          class="flex h-10 items-center rounded-lg bg-green-500 dark:bg-green-800 px-4 text-sm font-medium text-white transition-colors hover:bg-green-400 dark:hover:bg-green-600 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-blue-500 active:bg-blue-600 aria-disabled:cursor-not-allowed aria-disabled:opacity-50"
          type="submit"
        >
          Update
        </button>
      </div>
    </div>
  </form>
</div>
`;
